File: Amazing Bird/Classes/Sprites/ABGameCharacter.m
//
//  ABGameCharacter.m
//  Amazing Bird
//
//  Created by ChristianEnevoldsen on 25/02/14.
//  Copyright (c) 2014 ChristianEnevoldsen. All rights reserved.
//

#import "ABGameCharacter.h"
#import "ABGameScene.h"

@implementation ABGameCharacter {
	
}

- (instancetype)initWithTexture:(SKTexture *)texture position:(CGPoint)position
{
	self = [super initWithTexture:texture];
	if (self)
	{
		self.position = position;
		[self configurePhysicsBody];
	}
	return self;
}

- (void)configurePhysicsBody { /* up to subclass */}
- (void)collideWith:(SKPhysicsBody *)other { /* subclass */}

- (void)reset {
	[self removeAllActions];
	self.alpha = 1;
}

- (BOOL)applyDamage:(int)damage
{
	self.hitPoints -= damage;
	if (self.hitPoints < 0) {
		self.hitPoints = 0;
	}
	return self.hitPoints == 0;
}

// Acting
- (void)idle {

}

- (void)move {
}

- (void)performDeath {

}

- (void)revive {

}

- (void)update:(NSTimeInterval)delta {}

+ (void)loadSharedFrames { /* overriden by subclass */ }


- (void)animateWithFrames:(NSArray *)frames interval:(NSTimeInterval)interval withKey:(NSString *)key
{
	[self removeActionForKey:self.activeAnimationKey];
	[self runAction:[SKAction sequence:@[[SKAction animateWithTextures:frames timePerFrame:interval],
										 [SKAction runBlock:^{
		[self completedAnimationWithKey:key];
	}]]] withKey:key];
	self.activeAnimationKey = key;
}

- (void)completedAnimationWithKey:(NSString *)animKey { /* Overriden by subclass */ }

// Scene
- (void)addToScene:(ABGameScene *)scene atPosition:(CGPoint)position {
	[scene addNode:self atWorldLayer:ABWorldLayerGameCharacters];
	self.position = position;
}
- (ABGameScene *)characterScene {
	if ([self.scene isKindOfClass:[ABGameScene class]]) {
		return (ABGameScene *)self.scene;
	}
	return nil;
}


@end

File: Amazing Bird/Classes/Sprites/ABGameCharacter.swift
import SpriteKit

// MARK: - Enums

enum ABGameCharacterStatus {
    case idle, moving, dying, dead, reviving

    var isParalyzed: Bool {
        self == .dying || self == .dead || self == .reviving
    }
}

struct ABColliderType {
    static let bird:  UInt32 = 1
    static let cloud: UInt32 = 2
}

enum ABWorldLayer: Int {
    case background = 0
    case stars
    case gameCharacters
    case foreground

    static let count = 4
}

// MARK: - ABGameCharacter

class ABGameCharacter: SKSpriteNode {
    var hitPoints = 0
    var maxHitPoints = 0
    var movementSpeed: CGFloat = 0
    var status: ABGameCharacterStatus = .idle
    private(set) var activeAnimationKey: String?

    init(texture: SKTexture, position: CGPoint) {
        super.init(texture: texture, color: .clear, size: texture.size())
        self.position = position
        configurePhysicsBody()
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) not supported") }

    // MARK: - Overridable

    func configurePhysicsBody() {}
    func collideWith(_ body: SKPhysicsBody) {}

    func reset() {
        removeAllActions()
        alpha = 1
    }

    @discardableResult
    func applyDamage(_ damage: Int) -> Bool {
        hitPoints = max(0, hitPoints - damage)
        return hitPoints == 0
    }

    func idle() {}
    func move() {}
    func performDeath() {}
    func revive() {}
    func update(_ delta: TimeInterval) {}

    class func loadSharedFrames() {}

    // MARK: - Animation

    func animateWithFrames(_ frames: [SKTexture], interval: TimeInterval, key: String) {
        removeAction(forKey: activeAnimationKey ?? "")
        let animate = SKAction.animate(with: frames, timePerFrame: interval)
        let notify  = SKAction.run { [weak self] in self?.completedAnimation(key: key) }
        run(SKAction.sequence([animate, notify]), withKey: key)
        activeAnimationKey = key
    }

    func completedAnimation(key: String) {}

    // MARK: - Scene

    func addToScene(_ scene: ABGameScene, atPosition position: CGPoint) {
        scene.addNode(self, atWorldLayer: .gameCharacters)
        self.position = position
    }

    var characterScene: ABGameScene? { scene as? ABGameScene }
}
